plan/predicate_push_down.go

```diff
@@ -47,10 +48,225 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
 	return expr
 }
 
+// propagateConstant propagate constant values of equality predicates and inequality predicates in a condition.
+func propagateConstant(conditions []expression.Expression) []expression.Expression {
+	if len(conditions) == 0 {
+		return conditions
+	}
+	// Propagate constants in equality predicates.
+	// e.g. for condition: "a = b and b = c and c = a and a = 1";
+	// we propagate constant as the following step:
+	// first: "1 = b and b = c and c = 1 and a = 1";
+	// next:  "1 = b and 1 = c and c = 1 and a = 1";
+	// next:  "1 = b and 1 = c and 1 = 1 and a = 1";
+	// next:  "1 = b and 1 = c and a = 1";
+
+	// e.g for condition: "a = b and b = c and b = 2 and a = 1";
+	// we propagate constant as the following step:
+	// first: "a = 2 and 2 = c and b = 2 and a = 1";
+	// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+	// next:  "0"
+	isHandled := make([]bool, len(conditions))
+	type transitiveEqualityPredicate map[string]*expression.Constant // transitive equality predicates between one column and one constant
+	for {
+		equalities := make(transitiveEqualityPredicate, 0)
+		for i, getOneEquality := 0, false; i < len(conditions) && !getOneEquality; i++ {
+			if isHandled[i] {
+				continue
+			}
+			expr, ok := conditions[i].(*expression.ScalarFunction)
+			if !ok {
+				continue
+			}
+			// process the included OR conditions recursively to do the same for CNF item.
+			switch expr.FuncName.L {
+			case ast.OrOr:
+				expressions := expression.SplitDNFItems(conditions[i])
+				newExpression := make([]expression.Expression, 0)
+				for _, v := range expressions {
+					newExpression = append(newExpression, propagateConstant([]expression.Expression{v})...)
+				}
+				conditions[i] = expression.ComposeDNFCondition(newExpression)
+				isHandled[i] = true
+			case ast.AndAnd:
+				newExpression := propagateConstant(expression.SplitCNFItems(conditions[i]))
+				conditions[i] = expression.ComposeCNFCondition(newExpression)
+				isHandled[i] = true
+			case ast.EQ:
+				var (
+					col      *expression.Column
+					val      *expression.Constant
+					isColumn bool
+				)
+				left, ok1 := expr.Args[0].(*expression.Constant)
```

Comment:
rename `left`, `ok1`, `right`, `ok2` to `leftConst`, `leftIsConst`, `rightConst`, `rightIsConst`
and add lines
```
leftCol, leftIsCol := expr.Args[0].(*expression.Column)
rightCol, rightIsCol := expr.Args[1].(*expression.Column)
```
could be more readable.